File: src/MandelMap.cpp
// SPDX-FileCopyrightText: 2010-2025 Friedrich von Never <friedrich@fornever.me>
//
// SPDX-License-Identifier: MIT

#include "MandelMap.h"

#include <cassert>

using namespace std;

MandelMap::MandelMap(int width, int height)
{
    minMaxCached = false;
    this->width = width;
    this->height = height;
    data = new int[width * height];
}

MandelMap::~MandelMap()
{
    delete[] data;
}

int MandelMap::getWidth() const
{
    return width;
}

int MandelMap::getHeight() const
{
    return height;
}

void MandelMap::set(int x, int y, int value)
{
#ifdef _DEBUG
    assert(x < width && y < height && y * width + x < width * height);
#endif

    data[y * width + x] = value;
}

void MandelMap::draw(SDL_PixelFormat *pixelFormat, vector<Uint32> &pixels)
{
    minMaxCached = false;

    int maxIteration = getMax();
    for(int y = 0; y < height; ++y)
    {
        for(int x = 0; x < width; ++x)
        {
            float grade = (float)(get(x, y)) / maxIteration;
            auto &p_pixel = pixels[x + y * width];
            Uint8 R = grade * 256;
            Uint8 G = 128 + grade * 128;
            Uint8 B = 256 - grade * 256;
            auto color = SDL_MapRGB(pixelFormat, R, G, B);
            p_pixel = color;
        }
    }
    getMin();
}

int MandelMap::get(int x, int y) const
{
#ifdef _DEBUG
    assert(x < width && y < height && y * width + x < width * height);
#endif

    return data[y * width + x];
}

int MandelMap::getMin()
{
    if(!minMaxCached)
        cacheMinMax();

    return min;
}

int MandelMap::getMax()
{
    if(!minMaxCached)
        cacheMinMax();

    return max;
}

void MandelMap::cacheMinMax()
{
    min = max = data[0];
    for(int y = 0; y < height; ++y)
    {
        for(int x = 0; x < width; ++x)
        {
            if(get(x, y) < min)
                min = get(x, y);
            if(get(x, y) > max)
                max = get(x, y);
        }
    }
    minMaxCached = true;
}

File: src/mandel_map.rs
// SPDX-FileCopyrightText: 2010-2025 Friedrich von Never <friedrich@fornever.me>
//
// SPDX-License-Identifier: MIT

use sdl2::pixels::PixelFormat;

pub struct MandelMap {
    width: usize,
    height: usize,
    data: Vec<i32>,
    min: Option<i32>,
    max: Option<i32>,
}

impl MandelMap {
    pub fn new(width: usize, height: usize) -> Self {
        let mut data = Vec::with_capacity(width * height);
        data.resize(width * height, 0);
        Self {
            width,
            height,
            data,
            min: None,
            max: None,       
        }
    }
    
    fn set(&mut self, x: usize, y: usize, value: i32) {
        self.data[y * self.width + x] = value;   
    }
    
    fn draw(&self, pixel_format: PixelFormat, pixels: &mut Vec<u32>) {
        panic!("Not implemented");       
    }
    
    fn calc_min_max(&mut self) {
        panic!("Not implemented");       
    }
}
